lib/multiplexer/multiplexer.go

```diff
@@ -246,44 +256,146 @@ func (m *Mux) detectAndForward(conn net.Conn) {
+
+// detect finds out a type of the connection and returns wrapper that support PROXY protocol
+// We pass `casGetter` as function instead of certs themselves, so it could be called only if it's needed.
+func (m *Mux) detect(conn net.Conn) (*Conn, error) {
 	reader := bufio.NewReader(conn)
 
-	// the first attempt is to parse optional proxy
-	// protocol line that is injected by load balancers
-	// before actual protocol traffic flows.
-	// if the first attempt encounters proxy it
-	// goes to the second pass to do protocol detection
+	// Before actual protocol traffic flows, we try to parse optional PROXY protocol headers,
+	// that can be injected by load balancers or our own proxies. There can be multiple PROXY
+	// headers. After they are parsed, last pass does the actual protocol detection itself.
+	// We allow only one unsigned PROXY header from external sources, if it's enabled, and unlimited
+	// amount (within maxPROXYPasses) of signed headers from our own proxies, which take precedence.
 	var proxyLine *ProxyLine
-	for i := 0; i < 2; i++ {
+	for i := 0; i < maxPROXYPasses; i++ {
 		proto, err := detectProto(reader)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
 
 		switch proto {
 		case ProtoProxy:
-			if !enableProxyProtocol {
+			if !m.EnableExternalProxyProtocol {
 				return nil, trace.BadParameter("proxy protocol support is disabled")
 			}
+			// We allow only one unsigned proxy line
 			if proxyLine != nil {
 				return nil, trace.BadParameter("duplicate proxy line")
 			}
 			proxyLine, err = ReadProxyLine(reader)
 			if err != nil {
 				return nil, trace.Wrap(err)
 			}
+			// repeat the cycle to detect the protocol
 		case ProtoProxyV2:
-			if !enableProxyProtocol {
-				return nil, trace.BadParameter("proxy protocol support is disabled")
+			newProxyLine, err := ReadProxyLineV2(reader)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			verified := false
+			// If TLVs are empty we know it can't be signed, so we don't try to verify to avoid unnecessary load
+			if m.CertAuthoritiesGetter != nil && len(newProxyLine.TLVs) > 0 {
+				hostCACerts := [][]byte{}
+				certAuthorities, err := m.CertAuthoritiesGetter.GetCertAuthorities(m.context, types.HostCA, false)
+				if err != nil {
+					return nil, trace.Wrap(err, "could not get certificate authorities")
+				}
+
+				for _, ca := range certAuthorities {
+					caCerts := services.GetTLSCerts(ca)
+					hostCACerts = append(hostCACerts, caCerts...)
+				}
+				verified, err = newProxyLine.VerifySignature(hostCACerts, m.Clock)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			// If proxy line is signed and successfully verified, it's always accepted as current proxy line
+			if verified {
+				proxyLine = newProxyLine
+				continue
 			}
+
+			// If proxy line was signed but couldn't be verified it's ignored
+			if !verified && newProxyLine.isSigned() {
```

Comment:
I think we might want to take a harsher stance here and reject the connection entirely - unless we imagine there's a normal operating scenario where this is likely to be signed but it could fail verification.

My fear being that if this fails and only logs, the user is likely to experience much more confusing errors (e.g if using IP pinning, they may end up having their connections rejected for coming from the wrong IP, they'd have to know this log message was related to their issue to work this out).

This harsher stance would also complicate things since we could make `VerifySignature` simply return an error rather than `(boolean, error)`

Reply:
Yes, I agree. I was on a fence about this, and probably taking a harsher stance makes sense 👍 

> This harsher stance would also **complicate** things 

Did you mean "simplify"? 😅

Reply:
Urm, yes I did ! Might've reviewed this PR half asleep :')

Reply:
Changed it to the harsher stance logic.